docs/java-rest/high-level/document/reindex.asciidoc

```diff
@@ -0,0 +1,211 @@
+in the request body
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[reindex-request-conflicts]
+--------------------------------------------------
+<1> Set `proceed` on version conflict
+
+You can limit the documents by adding a type to the source or by adding a query.
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[reindex-request-typeOrQuery]
+--------------------------------------------------
+<1> Only copy `doc` type
+<2> Only copy documents which have field `user` set to `kimchy`
+
+It’s also possible to limit the number of processed documents by setting size.
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[reindex-request-size]
+--------------------------------------------------
+<1> Only copy 10 documents
+
+By default `_reindex` uses batches of 1000. You can change the batch size with `sourceBatchSize`.
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[reindex-request-sourceSize]
+--------------------------------------------------
+<1> Use batches of 100 documents
+
+Reindex can also use the ingest feature by specifying a `pipeline`.
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[reindex-request-pipeline]
+--------------------------------------------------
+<1> set pipeline to `my_pipeline`
+
+If you want a particular set of documents from the source index you’ll need to use sort. If possible, prefer a more
+selective query to size and sort.
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[reindex-request-sort]
+--------------------------------------------------
+<1> add descending sort to`field1`
+<2> add ascending sort to `field2`
+
+`ReindexRequest` also supports a `script` that modifies the document. It allows you to also change the document's
+metadata. The following example illustrates that.
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[reindex-request-script]
+--------------------------------------------------
+<1> `setScript` to bump the version of the source document
+
+`ReindexRequest` supports reindexing from a remote Elasticsearch cluster. When using a remote cluster the query should be
+specified inside the `RemoteInfo` object and not using `setSourceQuery`.
```

Comment:
I'd explicitly say that the query set by `setSourceQuery` is ignored. It might also be nice to explain why: the remote Elasticsearch may not understand queries built by the modern query builders. This works all the way back to Elasticsearch 0.90 and the query language has drifted a bit since then. When you reach to old versions, it is better to write the query by hand in json.